### ISSUE TITLE
Gateway API: serverNames should be unset if hostname not specified in Gateway config

### ISSUE THREAD

**Issue by ttreptow** [cilium/cilium#27099]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

My gateway looks like this:

```
apiVersion: gateway.networking.k8s.io/v1beta1
kind: Gateway
metadata:
  name: mygateway
spec:
  gatewayClassName: cilium
  listeners:
  - allowedRoutes:
      namespaces:
        from: Same
    name: gateway
    port: 443
    protocol: HTTPS
    tls:
      certificateRefs:
      - name: my-cert
      mode: Terminate

```

Note the lack of a `hostname`. 

The HTTPRoute looks like this:
```
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: test-route
  namespace: test
spec:
  parentRefs:
    - group: gateway.networking.k8s.io
      kind: Gateway
      name: mygateway
  hostnames:
    - myhost.example.com
  rules:
    - backendRefs:
        - name: test-svc
          port: 80
```

When I apply the HTTPRoute, I see errors in the cilium-envoy pod:

> [2023-07-26 16:22:32.308][1][warning][config] [external/envoy/source/common/config/grpc_subscription_impl.cc:128] gRPC config for type.googleapis.com/envoy.config.listener.v3.Listener rejected: Error adding/updating listener(s) test/cilium-gateway-mygateway/listener: error adding listener '127.0.0.1:19063': partial wildcards are not supported in "server_names"

The issue seems to be that the filterChainMatch field "serverNames" is a list with a single '*' element. The [envoy docs](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/listener/v3/listener_components.proto#config-listener-v3-filterchainmatch) state:

> Note that partial wildcards are not supported, and values like *w.example.com are invalid. The value * is also not supported, and server_names should be omitted instead.

### Cilium Version

1.14.rc.1

### Kernel Version

5.19.0-1028-aws

### Kubernetes Version

1.26

### Anything else?

The generated CiliumEnvoyConfig:
```
apiVersion: cilium.io/v2
kind: CiliumEnvoyConfig
metadata:
  creationTimestamp: "2023-07-26T17:14:56Z"
  generation: 2
  name: cilium-gateway-mygateway
  namespace: test
  ownerReferences:
  - apiVersion: gateway.networking.k8s.io/v1beta1
    kind: Gateway
    name: mygateway
    uid: f2d4facf-fa4b-4ec7-87ab-e6bd21d3703c
  resourceVersion: "9247016"
  uid: 064a87f5-ce87-45ab-99a5-16bf2cd03777
spec:
  backendServices:
  - name: test-svc
    namespace: test
    number:
    - "80"
  resources:
  - '@type': type.googleapis.com/envoy.config.listener.v3.Listener
    filterChains:
    - filterChainMatch:
        transportProtocol: raw_buffer
      filters:
      - name: envoy.filters.network.http_connection_manager
        typedConfig:
          '@type': type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          httpFilters:
          - name: envoy.filters.http.router
            typedConfig:
              '@type': type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
          rds:
            routeConfigName: listener-insecure
          statPrefix: listener-insecure
          upgradeConfigs:
          - upgradeType: websocket
          useRemoteAddress: true
    - filterChainMatch:
        serverNames:
        - '*'
        transportProtocol: tls
      filters:
      - name: envoy.filters.network.http_connection_manager
        typedConfig:
          '@type': type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          httpFilters:
          - name: envoy.filters.http.router
            typedConfig:
              '@type': type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
          rds:
            routeConfigName: listener-secure
          statPrefix: listener-secure
          upgradeConfigs:
          - upgradeType: websocket
          useRemoteAddress: true
      transportSocket:
        name: envoy.transport_sockets.tls
        typedConfig:
          '@type': type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.DownstreamTlsContext
          commonTlsContext:
            tlsCertificateSdsSecretConfigs:
            - name: cilium-secrets/test-my-cert
    listenerFilters:
    - name: envoy.filters.listener.tls_inspector
      typedConfig:
        '@type': type.googleapis.com/envoy.extensions.filters.listener.tls_inspector.v3.TlsInspector
    name: listener
    socketOptions:
    - description: Enable TCP keep-alive (default to enabled)
      intValue: "1"
      level: "1"
      name: "9"
      state: STATE_LISTENING
    - description: TCP keep-alive idle time (in seconds) (defaults to 10s)
      intValue: "10"
      level: "6"
      name: "4"
      state: STATE_LISTENING
    - description: TCP keep-alive probe intervals (in seconds) (defaults to 5s)
      intValue: "5"
      level: "6"
      name: "5"
      state: STATE_LISTENING
    - description: TCP keep-alive probe max failures.
      intValue: "10"
      level: "6"
      name: "6"
      state: STATE_LISTENING
  - '@type': type.googleapis.com/envoy.config.route.v3.RouteConfiguration
    name: listener-secure
    virtualHosts:
    - domains:
      - myhost.example.com
      - myhost.example.com:*
      name: myhost.example.com
      routes:
      - match:
          safeRegex:
            regex: (/.*)?$
        route:
          cluster: test/test-svc:80
          maxStreamDuration:
            maxStreamDuration: 0s
  - '@type': type.googleapis.com/envoy.config.cluster.v3.Cluster
    connectTimeout: 5s
    name: test/test-svc:80
    outlierDetection:
      splitExternalLocalOriginErrors: true
    type: EDS
    typedExtensionProtocolOptions:
      envoy.extensions.upstreams.http.v3.HttpProtocolOptions:
        '@type': type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions
        commonHttpProtocolOptions:
          idleTimeout: 60s
        useDownstreamProtocolConfig:
          http2ProtocolOptions: {}
  services:
  - listener: ""
    name: cilium-gateway-mygateway
    namespace: test

```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by youngnick** [comment]
Thanks for logging this @ttreptow, that seems like a bug.

I think that whoever looks at this should also consider adding a conformance test upstream to catch this case as well. Seems odd there isn't one already.

**Comment by sergeyshevch** [comment]
I can confirm the same issue on the released cilium 1.14. 

Workaround is to set exact domain/wildcard

**Comment by sergeyshevch** [comment]
I looked at the source code to fix this issue and found that it's already fixed in #27472. Thanks to @sayboras

Waiting for 1.14.2 to be released

**Comment by sayboras** [comment]
Thanks for confirming that this issue is fixed as part of #27472, which is currently backported to v1.14 branch in https://github.com/cilium/cilium/pull/27655. 

I am closing this issue, feel free to re-open if the issue is still happening once v1.14.2 is released.
